### ISSUE TITLE
Execution priority on sub dags

### ISSUE THREAD

**Issue by davidz-1** [dagster-io/dagster#4200]
## Use Case
We have a pipeline that processes data with using the `map` method, and looks roughly like this.
1. check_jobs_to_process
2. load_raw_data (much data)
3. preprocess_raw_data (heavily reduces data)
4. do_fancy_stuff
5. push_results

So solids 2-5 get mapped on the amount of jobs from 1 and get never collected at the end.
The actual issue with this is, that the current execution order first executes step 2 for all jobs and then 3, 4, 5 for each job. Which means at the moment when solid 2 is finished for all jobs, the amount of data in memory is pretty high, currently we do not have the possibility to add an `io_management`.
Since the data get's heavily reduced on solid 3 the moment in which the memory usage is high is actually pretty short.

## Ideas of Implementation
Add a configuration, or maybe a parameter to the `map` method to execute all sub dags in depth first instead of executing on width first across all mapped jobs.

---
#### Message from the maintainers:
Excited about this feature? Give it a :thumbsup:. We factor engagement into prioritization.

**Comment by alangenfeld** [comment]
This is a pretty buried feature
https://docs.dagster.io/_apidocs/execution#dagster.in_process_executor
but you can set `dagster/priotiy` tags on your solids to effect how the executors select the next to execute. I think you can use this to solve your problem. 

Let me know if it does not work as expected.

**Comment by yuhan** [comment]
This can be an example on the Pipeline Execution page https://github.com/dagster-io/dagster/issues/3964

**Comment by davidz-1** [comment]
@alangenfeld It worked just fine, I went the way to apply a negative priority to the data fetching steps. The more data is expected the smaller the priority gets.
Thank you for the fast reply.

**Comment by AntonFriberg** [comment]
Hi! I think I have a related issue.

In my case, I wanted to manage memory issues in a simple ETL pipeline? I extract from Postgres, transform into Elasticsearch dictionaries, load it into Elasticsearch. The issue is that the extracted table from Postgres might be very large so I would like to chunk the outputs.

I have tried doing this with the DynamicOutput solution but it seems that it yields all the steps before continuing down the pipeline. I did manage to make the transform and load steps in the correct order by utilizing priority tags but it does not really matter if the extract step yields all the outputs before continuing.

I have also tried recreating the downstream solids into a single composite solid, changing the IOManager into a pickle filesystem one, and tweaking the priority of the tasks. So far I have not been successful and the memory usage of all the different implementations has been about the same. If you see in my screenshot it seems all rows are yielded before continuing downstream in the graph. I would like to see yield -> load, yield -> load, and so on.

TLDR: How do I chunk outputs from a solid and let the downstream solids run in order for each output to limit the memory usage of the pipeline?

![image](https://user-images.githubusercontent.com/14290517/121019290-92b51300-c79f-11eb-82b8-3f7ec60d76ca.png)

![Screenshot 2021-06-07 at 13-23-50 Pipeline featuredb2elastic](https://user-images.githubusercontent.com/14290517/121019330-9c3e7b00-c79f-11eb-8477-f774f75348f7.png)

**Comment by alangenfeld** [comment]
The current semantics of the dependencies in the system is that the upstream solid both returned the required output and completed successfully, so I don't think it is possible today to pull off the approach in the way you've described. 

How are you chunking up your table in to Dataframes? My recommendation would be to have the solid thats using `DynamicOutput` to instead of returning the `Dataframe`, return the slicing information that can be used to create the `Dataframe` chunk independently.

**Comment by AntonFriberg** [comment]
I am doing the following to chunk the Dataframe

```python
@solid(required_resource_keys={"feature_database"}, output_defs=[DynamicOutputDefinition(pd.DataFrame)], tags={"dagster/priority": "-1"})
def extract_featuredb_dataframes(context, query: str):
    """Fetch data from the database and give as DataFrames."""
    with context.resources.feature_database.connect() as connection:
        context.log.info(f"Executing query: {query}")
        dataframe_chunks = pd.read_sql(query, connection, chunksize=100_000)
        for part, dataframe in enumerate(dataframe_chunks):
            context.log.info(f"Yielding {len(dataframe)} rows")
            yield DynamicOutput(value=dataframe, mapping_key=f"df_part_{part}")
```

I do not really understand how the slicing information would help. I would think that simply pushes the problem down the pipeline to the next solid (transformation solid in my case). I thought that perhaps it would help to use an IO manager to push the result to file between the solids, but that increased the memory usage rather than mitigate the issue.

If pointers are simply passed down the pipeline then it would defeat the purpose behind Dagster from my point of view. I value the type checking that is done between the steps.

If we had support for looping through parts of a pipeline with yielded outputs from an upstream solid I believe the use case of Dagster would greatly improve since we would be able to do much bigger tasks without resorting to complex setups such as Spark och Dask which are not able to be run locally on a developer workstation.

**Comment by AntonFriberg** [comment]
Would this issue perhaps be solved with the changes coming in v0.12? https://github.com/dagster-io/dagster/discussions/2902#discussioncomment-821227

**Comment by alangenfeld** [comment]
> I do not really understand how the slicing information would help
yea my recommendation does not really make sense for this set-up, its only reasonable if you are calculating the chunks ranges explicitly. 

> Would this issue perhaps be solved with the changes coming in v0.12
No those do not effect execution semantics. 

> I thought that perhaps it would help to use an IO manager to push the result to file between the solids, but that increased the memory usage rather than mitigate the issue.

I think this approach may actually be within close reach. We currently hold on to all step outputs in memory to make them available for `hooks` , but we could either add an explicit disable or not collect them if there are no hook defs. This in addition to ensuring there are no unintended references should get this approach working.

**Comment by AntonFriberg** [comment]
Thanks! Utilizing an IO manager would work as long as the data fits in the configured disk. 

I am not aware of how the DynamicOutputs builds the execution graph but I need to express my concern about how limiting it seems to not be able to start transforming parts of the data until all the data has been extracted.

**Comment by alangenfeld** [comment]
> express my concern about how limiting it seems to not be able to start transforming parts of the data until all the data has been extracted

Makes sense, this is a much bigger change for the system so will track in a separate issue.
